numba/core/overload_glue.py

```diff
@@ -136,6 +138,24 @@ def body(tyctx):
                 if self._TYPER is None:
                     raise errors.InternalError(msg)
                 typing = self._TYPER(tyctx)
+                # Try and bind the args to the typer sig, if this fails it's
+                # probably because the arg count doesn't match template sig.
+                generic = getattr(typing, 'generic', None)
+                if generic is not None:
+                    if isinstance(typing, CallableTemplate):
+                        generic_inst = generic()
+                        try:
+                            match_sig = inspect.signature(generic_inst)
+                            match_sig.bind(*ol_args, **ol_kwargs)
+                        except TypeError as e:
+                            # ctor or bind failed, raise, if there's a
+                            # ValueError then there's likely unrecoverable
+                            # problems
+                            raise errors.TypingError(str(e)) from e
```

Comment:
why can't this be part of `CallableTemplate.apply`?

Reply:
It can and probably should be, so long as it's generic enough?! Done in 810b6bf